mautic/mautic
# Copy button for segment filters

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [x]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | 



#### Description:

This new button allows users to copy segment groups to speed up segment filter building. The copy button is hidden for filters with the AND glue. It allows to copy only filters with the OR glue including all filters with the AND glue attached to it = filter group.

For reference, this is how the segment filters look before this change:

![Screenshot 2024-10-18 at 17 36 18](https://github.com/user-attachments/assets/23ae94b8-835b-443d-a031-8e90b6870f6d)


And this is after:
![Screenshot 2024-10-18 at 17 35 55](https://github.com/user-attachments/assets/fd8f51b7-df0b-4dda-b019-39b921873246)


The changes aren't only about adding the button. The delete button was moved to the top right corner of the filter header and the glue (and/or selection) was moved to the top left column. The filter value is now wider and we can add as many buttons as we want without affecting the filter value.

From the dev point of view, I did a small refactoring to DRY the code out a little.

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Test that the copy button works well in all cases:
3. When a new filter is added. Especially segment filter or tag filters which have multiselect as a filter value.
4. When you save the segment, you must be able to copy the filters as well.
5. When moving the filters around (changing filter order) then all the functionality must work as before and the copy button must work as well.
6. Test that the dynamic web content filters still work as before. Those share some code with segment filters.

#### Other areas of Mautic that may be affected by the change:
1. Dynamic web content

#### List deprecations along with the new alternative:
1. None



[//]: # ( As applicable: )
#### List of areas covered by the unit and/or functional tests:
1. There are no tests because the changes were made in the JS and PHPUNIT cannot test JS.
